[알트] 로또 미션 리뷰요청합니다.

KS-KIM opened the pull request:
안녕하세요. 많은 피드백 부탁드립니다!

lalize committed:
docs: 기능 목록 추가

lalize committed:
chore: 불필요한 파일 삭제

lalize committed:
feat: LottoMoney 클래스 추가

lalize committed:
feat: LottoNumber 클래스 추가

lalize committed:
refactor: 하드코딩 값 상수화 및 로또 수 비교 메서드 구현

lalize committed:
feat: LottoNumbers 클래스 추가

lalize committed:
feat: 로또 당첨 순위를 구하는 기능 추가

등수를 결정하기 위한 목적으로 LottoRank 열거형 추가
당첨번호를 저장하는 WinningLottoNumbers 추가

lalize committed:
refactor: 전체적인 네이밍 수정 및 테스트를 위한 메소드 추가

lalize committed:
feat: Lottos 클래스 추가

lalize committed:
feat: LottoStatistics 클래스 추가

lalize committed:
feat: RandomLottoGenerator 클래스 추가

KS-KIM committed:
refactor: 컨벤션에 적합하게 포매팅 및 코드 가독성 개선

모든 클래스 헤더 주석 추가
RandomLottoGenerator 반복문 가독성 개선
WinningLotto 상수 분리

KS-KIM committed:
feat: MatchCount 클래스 추가

KS-KIM committed:
refactor: 순위 결정 방식 개선

각 순위가 자신을 판별하는 로직을 가지도록 변경
랜덤 로또를 생성하는 인터페이스 변경

KS-KIM committed:
feat: 통계 결과를 출력하기 위한 TextUtil 클래스 추가

KS-KIM committed:
feat: InputView, OutputView 클래스 추가

KS-KIM committed:
refactor: LottoApplication 실행로직 분리

jihan805 commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -0,0 +1,48 @@
+		return new WinningLotto(lottoNumbers, bonusNumber);
+	}
+
+	public void play() {
```
LottoGame 클래스 추출 👍 
해당 클래스에 Output과 연관된 로직들이 보이는데, 이를 분리해보면 어떨까요?

MVC 패턴으로 구현하기!

KS-KIM replied:
기존에 `LottoApplication`에서 수행하는 입력 로직을 `LottoGame` 클래스로 이동시켜서 `LottGame` 클래스가 입출력과 도메인 로직 사이의 중재 역할을 수행하도록 변경했습니다!

또한, 입출력 과정에서 생기는 클래스를 필드로 두는 것은 불필요한 행위라고 생각되어 모두 `lottoGame`의 `run()` 메서드 지역 변수로 이동시켰습니다!

자동으로 로또 번호를 생성하는 `lottoTicketFactory` 클래스는 `run()` 메서드가 여러번 호출되어도 기존 호출에 영향 없이 동일한 동작을 수행하기에 `LottoGame` 클래스의 멤버 변수로 만들어 보았습니다.

jihan805 commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -0,0 +1,48 @@
+	public void play() {
+		OutputView.printLottos(lottos);
+		MatchResult matchResult = lottos.matchAll(winningLotto);
+		long totalProfits = matchResult.calculateTotalProfits(money);
```
사용하지 않는 변수 제거.

KS-KIM replied:
불필요한 변수를 제거했습니다!

jihan805 commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -0,0 +1,48 @@
+		return randomLottosFactory.generate(money.calculateBuyCount());
+	}
+
+	public WinningLotto generateWinningLotto(String winningLotto, int number) {
```
접근제어자가 public 일 필요가 있을까요?

KS-KIM replied:
private 접근 제어자로 변경했습니다!

jihan805 commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,62 @@
+
+	public long calculateTotalProfits(Money money) {
+		long totalWinning = calculateTotalWinnings();
+		return totalWinning * 100 / money.get();
```
상수 사용~

jihan805 commented on src/test/java/lotto/domain/MatchResultTest.java:
```diff
@@ -0,0 +1,27 @@
+		statistics.put(LottoRank.FIFTH, 2L);
+		statistics.put(LottoRank.MISS, 5L);
+		assertThat(new MatchResult(statistics)).isInstanceOf(MatchResult.class);
+	}
```
수익률 등에 대한 테스트가 안보이는 것 같아요!!

KS-KIM replied:
빠진 테스트에 대한 부분을 보완했습니다!

jihan805 commented on src/main/java/lotto/domain/LottoGeneratable.java:
```diff
@@ -0,0 +1,12 @@
+ * @author K.S.KIM
+ * @since 2020/02/19
+ */
+public interface LottoGeneratable {
```
인터페이스 추출했네요!
다음단계에서 잘 활용하면 좋을 것 같아요!

jihan805 reviewed (changes requested):
안녕하세요~! 전체적으로 구현 정말 잘하셨네요 👍 
간단한 피드백 확인해주세요 :)

KS-KIM committed:
refactor: 총 수익률을 구할 때 사용하는 상수 분리

KS-KIM committed:
refactor: 일급 컬렉션 Lotto를 불변 리스트로 생성하도록 변경

KS-KIM committed:
refactor: 로또 숫자 자동 생성방식 변경

생성할 로또의 개수를 원시값으로 받아오지 않고 돈을 입력받도록 변경
매번 숫자 리스트를 섞지 않고 범위 내의 랜덤 수를 생성하도록 변경
로또 숫자 자동 생성기에서 로또 수의 범위를 참조할 수 있도록 public 접근 제어자로 변경

KS-KIM committed:
refactor: 입출력과 도메인 사이의 흐름을 제어하는 LottoGame 개선

입출력을 LottoApplication이 아닌 LottoGame에서 받도록 변경
게임 수행 중 발생한 모든 exception은 LottoApplication이 제어한다.
catch된 exception은 별도의 복구 작업을 수행하지 않으므로 메시지만 출력한 후 종료한다.

KS-KIM committed:
refactor: LottoGame 생성자에서 null 객체가 들어왔을때 예외처리 추가

KS-KIM committed:
refactor: Lottos 클래스의 이름을 LottoTicket으로 변경

KS-KIM committed:
style: 누락된 클래스 주석 추가

KS-KIM committed:
refactor: Lottos 네이밍을 가진 클래스와 변수의 이름을 LottoTicket으로 변경

기존에 변경한 Lottos -> LottoTicket 클래스와 네이밍 일관성을 지키기 위함

KS-KIM committed:
refactor: 로또 번호 일치 여부를 확인하는 메서드 개선

디미터 법칙 준수

KS-KIM committed:
refactor: 로또 출력 메시지를 생성하는 방식 변경

format 문자열 제거 및 stream 활용 극대화

KS-KIM committed:
refactor: LottoNumber 불필요한 메서드 제거 및 테스트 케이스 추가

KS-KIM committed:
style: 숫자에 1,000 단위로 underscore 적용

KS-KIM committed:
refactor: Lottos -> LottoTicket으로 네이밍 변경

KS-KIM committed:
refactor: 생성된 모든 로또를 한번에 출력하도록 변경

문자열 출력 성능을 향상하기 위한 조치

KS-KIM committed:
test: 테스트 케이스 추가 및 변수명을 영어로 변경

kingbbode commented on src/main/java/lotto/domain/AutoLottoTicketFactory.java:
```diff
@@ -0,0 +1,34 @@
+	@Override
+	public LottoTicket generate(Money money) {
+		List<Lotto> lottos = new ArrayList<>();
+		for (long count = money.calculateBuyCount(); count > 0; --count) {
```
로또 가격에 대한 정보를 `Money` 가 알고 있는 것이 맞을까요?

금액으로부터 횟수를 구하는 것 대신, 가지고 있는 금액을 차감한다면 더 역할이 명확한 분리가 될 것 같아요 :)

KS-KIM replied:
말씀하신대로 횟수를 받아와 반복을 돌리는 것 보다는 구입 금액을 1,000원씩 사용하면서 잔액이 남아있는지 확인하는 것이 더 자연스러운 흐름같네요! 해당 부분 피드백 반영하였습니다 :)

kingbbode commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,70 @@
+	public static final int MIN_VALUE = 1;
+	public static final int MAX_VAULE = 45;
+	private static final String INVALID_NUMBER_MESSAGE = "생성할 수 없는 수입니다.";
+	private static final Map<Integer, LottoNumber> CACHE = new HashMap<>();
```
`CACHE` 라는 추상적인 네이밍보다는 명확한 네이밍을 해주는 것은 어떨까요?

KS-KIM replied:
`LottoNumber`, `MatchCount`클래스의 `CACHE`가 어떤 값을 캐싱하는지 표현하기 위해 각각 `LOTTO_NUMBER_CACHE`, `MATCH_COUNT_CACHE`로 변경했습니다.

kingbbode commented on src/test/java/lotto/domain/LottoRankTest.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	@Test
+	@DisplayName("로또 2등을 정상적으로 반환한 경우")
```
정상적이라는 범위가 무엇일까요?

테스트 설명은 설계 의도를 명확히 하는 것이 좋습니다.

`번호가 5개 일치하고, 보너스 번호가 일치하는 경우 2등이다`

KS-KIM replied:
테스트 DisplayName을 전반적으로 다시 점검하였습니다.

kingbbode reviewed (changes requested):
안녕하세요!
역할 및 책임에 따라 코드가 잘 작성된 코드네요 :) 👍 
몇 가지 코멘트 작성해두었습니다.
확인 부탁드려요.

KS-KIM committed:
refactor: LottoNumber 클래스에 포함된 변수명이 의도를 명확히 드러내도록 변경

CACHE라는 명칭이 어떤 값을 캐싱하는지 표현하기 위해 LOTTO_NUMBER_CACHE로 변경
MAX_VALUE 오타 수정

KS-KIM committed:
refactor: Money 클래스 명칭을 LottoPurchaseMoney로 변경

로또 게임에서 Money의 사용 목적을 보다 명확히 드러내기 위해 이름을 LottoPurchaseMoney로 변경

KS-KIM committed:
refactor: 로또 구입액 단위로 나누어 떨어지는지 여부를 확인하는 메서드명 변경

KS-KIM committed:
refactor: MatchCount의 CACHE 변수명 변경

CACHE가 어떤 값을 캐싱하는지 명확히 드러내기 위해 이름을 MATCH_COUNT_CACHE로 변경

KS-KIM committed:
refactor: 수익률 계산 방식 변경

KS-KIM committed:
refactor: 로또 통계결과를 만드는 스트림 수정

KS-KIM committed:
test: 테스트 display 메시지 변경

KS-KIM committed:
refactor: 로또 구입 방식 변경

구입 가능 횟수를 가져와 반복문을 돌리지 않고, 구입 금액을 1,000원씩 차감하면서 소모하도록 변경

kingbbode commented on src/main/java/lotto/domain/LottoPurchaseMoney.java:
```diff
@@ -11,35 +11,43 @@
 
-	public long calculateBuyCount() {
-		return lottoMoney / MONEY_UNIT;
+	public void pay() {
```
`LottoPurchaseMoney` 는 금액 관리에만 신경쓰고 금액을 지불을 요구하는 쪽에서 금액을 넘겨주면 더 유연하겠네요.
`LottoPurchaseMoney` 는 로또와 금액 사이의 모호한 책임을 가지고 있어요.
로또 금액은 로또를 발급하는 쪽에서 알고 있는 것이 더 자연스러운 역할관계 일 것 같아요 :)

kingbbode reviewed (approved):
피드백 반영 잘 해주셔서 감사합니다! 👍 
다음 단계도 잘 부탁드려요!
추가 코멘트도 확인 부탁드려요 :)
